src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs

```diff
@@ -56,9 +56,46 @@ public void Start(ApplicationData applicationData)
             top.Add(menu);
 
             var gridHeaders = applicationData.DataTable.DataColumns.Select((c) => c.Label).ToList();
-            // We add one as the offset here to line it up with the data.
-            win.Add(new Label(GetPaddedString(gridHeaders, top.Frame.Width - 3, offset: 1)));
-            
+            var columnWidths = new int[gridHeaders.Count];
+            for (int i = 0; i < gridHeaders.Count; i++)
+            {
+                columnWidths[i] = gridHeaders[i].Length;
+            }
+
+            // calculate the width of each column based on longest string in each column for each row
+            foreach (var row in applicationData.DataTable.Data)
+            {
+                int index = 0;
+                // use just the first 20 rows to calculate in case there is a large dataset
+                foreach (var col in row.Values.Take(20))
+                {
+                    var len = col.Value.DisplayValue.Length;
+                    if (len > columnWidths[index])
+                    {
+                        columnWidths[index] = len;
+                    }
+                    
+                    index++;
+                }
+            }
+
+            // If we have PassThru, then we want to make them selectable. If we make them selectable,
+            // they have a 8 character addition of a checkbox ("     [ ]") that we have to factor in.
+            int offset = applicationData.PassThru ? 8 : 4;
+
+            // if the total width is wider than the usable width, remove 1 from widest column until it fits
+            // the gui loses 3 chars on the left and 2 chars on the right
+            int usableWidth = top.Frame.Width - 3 - columnWidths.Length - offset - 2;
+            int columnWidthsSum = columnWidths.Sum();
+            while (columnWidthsSum >= usableWidth)
+            {
+                int maxWidth = columnWidths.Max();
+                int maxIndex = columnWidths.ToList().IndexOf(maxWidth);
```

Comment:
You turn this array into a list every time in that while loop and you calculate the max right above it (O(N) operation).

I think this can be optimized if you get the max yourself by using a loop. Then you can already have the index instead of using `ToList().IndexOf(..)`

Reply:
I can have a single loop to get rid of the Max() call, too.